gateway/loaders/zookeeper/zookeeper.go

```diff
@@ -0,0 +1,343 @@
+			z.config.Log.Info().Msg("Creating path: " + TargetPath)
+			if err := CreateFullPath(zClient.conn, TargetPath, zk.WorldACL(zk.PermAll)); err != nil {
+				return nil, err
+			}
+		}
+		if exists, _, _ := zClient.conn.Exists(CleanPath(RequestPath)); !exists {
+			z.config.Log.Info().Msg("Creating path: " + RequestPath)
+			if err := CreateFullPath(zClient.conn, RequestPath, zk.WorldACL(zk.PermAll)); err != nil {
+				return nil, err
+			}
+		}
+		return configs, nil
+	case nil:
+		z.config.Log.Info().Msg("Succesfully retreived zookeeper configuration")
+	default:
+		return nil, err
+	}
+
+	configs, err := z.zookeeperToTargets(data)
+
+	if err := target.Validate(configs); err != nil {
+		return nil, fmt.Errorf("configuration in zookeeper is invalid: %v", err)
+	}
+
+	return configs, nil
+}
+
+func (z *ZookeeperTargetLoader) Start() error {
+	_, err := z.GetConfiguration() // make sure there are no errors at startup
+	return err
+}
+
+func (z *ZookeeperTargetLoader) WatchConfiguration(targetChan chan<- *connections.TargetConnectionControl) error {
+	for {
+		targetConfig, err := z.GetConfiguration()
+		if err != nil {
+			z.config.Log.Error().Err(err).Msgf("Unable to get target configuration.")
+		} else {
+			controlMsg := new(connections.TargetConnectionControl)
+			if z.last != nil {
+				for targetName := range z.last.Target {
+					_, exists := targetConfig.Target[targetName]
+					if !exists {
+						controlMsg.Remove = append(controlMsg.Remove, targetName)
+					}
+				}
+			}
+			controlMsg.Insert = targetConfig
+			z.last = targetConfig
+
+			targetChan <- controlMsg
+		}
+		time.Sleep(z.interval)
+	}
+}
+
+func (z *ZookeeperClient) getNodeValue(nodePath string) ([]byte, error) {
+	data, _, err := z.conn.Get(nodePath)
+	if err != nil {
+		log.Log.Error(err, "zNode get failed", "path", nodePath)
+		return nil, err
+	}
+	return data, nil
+}
+
+func CleanPath(path string) string {
+	return "/" + strings.Trim(path, "/")
+}
+
+func (z *ZookeeperClient) ConnectToZookeeper(hosts []string, timeout time.Duration) (*zk.Conn, error) {
+	newConn, eventChan, err := zk.Connect(hosts, timeout)
+	if err != nil {
+		return nil, err
+	}
+	go z.zookeeperEventHandler(eventChan)
+	return newConn, nil
+}
+
+func (z *ZookeeperClient) zookeeperEventHandler(zkEventChan <-chan zk.Event) {
+	var disconnectedCount int
+	for event := range zkEventChan {
+		switch event.State {
+		case zk.StateDisconnected:
+			disconnectedCount++
+		case zk.StateHasSession:
+			disconnectedCount = 0
+		}
+	}
+}
+
+// TODO implement more options for zookeeper client connection
+func NewZookeeperClient(hosts []string, timeout time.Duration) (*ZookeeperClient, error) {
+	DefaultACLPerms := zk.PermAll
+	z := &ZookeeperClient{
+		acl: zk.WorldACL(int32(DefaultACLPerms)),
+	}
+	zkconn, err := z.ConnectToZookeeper(hosts, timeout)
+	if err != nil {
+		return nil, err
+	}
+	z.conn = zkconn
+
+	return z, nil
+}
+
+func CreateParentPath(conn *zk.Conn, path string, acl []zk.ACL) error {
```

Comment:
nit: we're reusing some Zookeeper helpers, those might be moved to a zookeeper utils module to avoid code duplication.